catalog/spatial/csw/spatial-csw-endpoint/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/endpoint/CswSubscriptionEndpoint.java

```diff
@@ -0,0 +1,546 @@
+        QueryRequest query = queryFactory.getQuery(request);
+
+        return new CswSubscription(mimeTypeTransformerManager, request, query);
+    }
+
+    public synchronized String addOrUpdateSubscription(GetRecordsType request,
+            boolean persistSubscription) throws CswException {
+        String methodName = "createSubscription";
+        if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug("ENTERING: {}    (persistSubscription = {})", methodName);
+        }
+
+        if (request.getResponseHandler() == null || request.getResponseHandler()
+                .isEmpty() || StringUtils.isEmpty(request.getResponseHandler()
+                .get(0))) {
+            throw new CswException(
+                    "Unable to create subscription because deliveryMethodUrl is null or empty");
+        }
+
+        String deliveryMethodUrl = request.getResponseHandler()
+                .get(0);
+
+        String subscriptionUuid = getSubscriptionUuid(request.getRequestId());
+        if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug("subscriptionUuid = {}", subscriptionUuid);
+        }
+        request.setRequestId(subscriptionUuid);
+
+        Dictionary<String, String> props = new Hashtable<>();
+        props.put("subscription-id", subscriptionUuid);
+
+        // If this subscription already exists, then delete it and re-add it
+        // to registry
+        if (registeredSubscriptions.containsKey(subscriptionUuid)) {
+            LOGGER.debug("Delete existing subscription {} for re-creation", subscriptionUuid);
+            deleteSubscription(subscriptionUuid);
+        }
+        CswSubscription sub = createSubscription(request);
+        LOGGER.debug("Registering Subscription");
+        ServiceRegistration serviceRegistration =
+                context.registerService(Subscription.class.getName(), sub, props);
+
+        if (serviceRegistration != null) {
+            LOGGER.debug("Subscription registered with bundle ID = {} ",
+                    serviceRegistration.getReference()
+                            .getBundle()
+                            .getBundleId());
+            registeredSubscriptions.put(subscriptionUuid, serviceRegistration);
+            // Pass in client-provided subscriptionId vs. subscription UUID because
+            // the filter XML to be persisted for this subscription will be used to
+            // restore this subscription and should consist of the exact values the
+            // client originally provided.
+            if (persistSubscription) {
+                persistSubscription(sub, deliveryMethodUrl, subscriptionUuid);
+            }
+        } else {
+            LOGGER.debug("Subscription registration failed");
+        }
+
+        LOGGER.debug("EXITING: {}", methodName);
+        return subscriptionUuid;
+
+    }
+
+    private synchronized boolean deleteSubscription(String subscriptionId) throws CswException {
```

Comment:
We may want to revisit the way we do synchronization in this class. We use a `ConcurrentHashMap` and also synchronize on the object itself. Would it be possible to use the `ConcurrentHashMap`'s special methods to avoid object level synchronization? If not, then maybe we should just use a plain `HashMap` and use it as our monitor and only synchronize on the map of synchronizing on two different objects.